GetStream/stream-chat-swift
# [CIS-1522] Fix Memory Leaks when opening and closing channels

### 🔗 Issue Link
CIS-1522

### 🎯 Goal
The main goal was to fix the memory leaks around opening and closing channels which was mostly caused by the `GalleryVC`, but an overall cleanup of the memory leaks on the SDK was done as well.

### 🛠 Implementation
Break reference cycles in multiple places of the codebase.

### 📝 Changes
- Make sure to invalidate the `URLSession` in the `URLWebSocketEngine`. According to Apple, the `URLSession.delegate` is strongly retained and we need to make sure we call `invalidateAndCancel()` to clean up the `URLSession`.
- Fix Ref Cycle in `GalleryVC`.
- Fix multiple ref cycles in  Updaters and Middlewares.
- Correctly unsubscribe from video player observers.
- Make sure to destroy the `ChatClient` in the `LoginViewController` so that is easier to analyse memory leaks
- Improve the feedback from the connection banner of the DemoApp

### 🎨 Changes
These are the instances that are retained after playing with the application and then coming back to the login screen before the fixes and after the fixes.
| Before | After |
| --- | --- |
| <img width="179" alt="image" src="https://user-images.githubusercontent.com/12814114/155738788-a21c5a4e-633c-4cd3-9da9-c1640e15c35d.png"> | <img width="173" alt="image" src="https://user-images.githubusercontent.com/12814114/155736499-064e6411-0df8-48e6-a3b7-eac5af992385.png"> |

### 🧪 Testing
- Run the DemoApp
- Select a User
- Open a Channel
- Scroll and open some images and videos
- Close a Channel
- Repeat the last 3 steps multiple times
- Logout
- Open the Memory Graph and observe the live instances

### ☑️ Contributor Checklist


### 🎉 GIF

<img src="https://media.giphy.com/media/VeSvZhPrqgZxx2KpOA/giphy.gif" width="50%"/>
